gcp/modules/couchdb/recover.tf

```diff
@@ -0,0 +1,83 @@
+# This resource is needed to restore PVs that were previously claimed by CouchDB
+# We need it to avoid data loss during forceful cluster re-creation
+# https://issues.gpii.net/browse/GPII-3493
+
+resource "null_resource" "couchdb_recover" {
+  triggers = {
+    nonce = "${var.nonce}"
+  }
+
+  provisioner "local-exec" {
+    command = <<EOF
+      for DISK in $(gcloud compute disks list --filter description:couchdb --format json | jq -c .[]); do
+        PD_NAME=$(echo $DISK | jq -r .name)
+        PD_DESC=$(echo $DISK | jq -r .description)
+        PV_NAME=$(echo $PD_DESC | jq -r '.["kubernetes.io/created-for/pv/name"]')
+        PVC_NAME=$(echo $PD_DESC | jq -r '.["kubernetes.io/created-for/pvc/name"]')
+        if [ "$(kubectl -n gpii get pvc $PVC_NAME -o json 2>/dev/null | jq -r .metadata.name)" != "$PVC_NAME" ]; then
+          jq -n \
+            --arg pd_name "$PD_NAME" \
+            --arg pv_name "$PV_NAME" \
+            --arg pvc_name "$PVC_NAME" \
+          '{
+            "apiVersion": "v1",
+            "items": [
+              {
+                "apiVersion": "v1",
+                "kind": "Namespace",
+                "metadata": {
+                  "name": "${var.release_namespace}",
+                }
+              },
+              {
+                "apiVersion": "v1",
+                "kind": "PersistentVolume",
+                "metadata": {
+                  "name": $pv_name
+                },
+                "spec": {
+                  "accessModes": [
+                    "ReadWriteOnce"
+                  ],
+                  "capacity": {
+                    "storage": "${var.pv_capacity}"
+                  },
+                  "gcePersistentDisk": {
+                    "fsType": "ext4",
```

Comment:
AIUI, the motivation for this custom code to handle PVs/PVCs (rather than managing PVs directly and completely with Terraform) is to let Google/Kubernetes worry about as much of the process as possible. So, it's weird to me that we have an entire PV/PVC definition here using kubectl -- is using kubectl better than using Terraform resources?

Relatedly, I worry about a couple parameters that we "let" Google/Kubernetes specify (like `fsType`) that we are then forced to shadow here. What happens if Google/Kubernetes change the default fsType for GCE Disks? How would we know to change this file? Could this scenario lead to data loss?

Reply:
Current implementation is the simplest way to recover PV and PVC objects from existing PDs that were previously managed by K8s.
Solution with managing all PD, PV and PVC resources in Terraform may seem better organized at first look, but it has complications with naming (hard to make sure that `couchdb-0` pod always uses PD with the same name), resource distribution across cluster zones and scaling.
So team decided to go for simpler approach.

And yes, there are still lots of unanswered questions. I don't think there is a way to specify filesystem type for [GCE PD](https://kubernetes.io/docs/concepts/storage/storage-classes/#gce-pd) at the moment, and I don't know what happens when they introduce it or change default filesystem from ext4 (but we'll surely notice that before it reaches prd environment). We recently discovered that even resizing existing PVs is pretty [complicated task](https://issues.gpii.net/browse/GPII-3532).

Do you think we should revisit current implementation sometime in near future?